reference-implementation/__tests__/composition.js

```diff
@@ -0,0 +1,367 @@
+'use strict';
+const { URL } = require('url');
+const { parseFromString } = require('../lib/parser.js');
+const { appendMap } = require('../lib/composer.js');
+
+const mapBaseURL = new URL('https://example.com/app/index.html');
+
+function composeMaps(mapLikes) {
+  if (!Array.isArray(mapLikes) || mapLikes.length < 1) {
+    throw new Error('composeMaps must be given a non-empty array of mapLikes');
+  }
+  let map = parseFromString(JSON.stringify(mapLikes.shift()), mapBaseURL);
+  for (const mapLike of mapLikes) {
+    const newMap = parseFromString(JSON.stringify(mapLike), mapBaseURL);
+    map = appendMap(map, newMap);
+  }
+  return JSON.parse(JSON.stringify(map));
+}
+
+describe('Composition', () => {
+  it('should compose with the empty map on the left', () => {
+    const map = parseFromString(JSON.stringify({
+      imports: { 'https://a/': ['https://b/'] },
+      scopes: {
+        'https://c/': { 'https://d/': ['https://e/'] }
+      }
+    }), mapBaseURL);
+
+    const resultMap = appendMap(parseFromString('{}', mapBaseURL), map);
+
+    expect(resultMap).toStrictEqual(map);
+    expect(resultMap.imports).not.toBe(map.imports);
+    Object.entries(resultMap.imports).forEach(([k, v]) => {
+      expect(v).not.toBe(map.imports[k]);
+    });
+    expect(resultMap.scopes).not.toBe(map.scopes);
+    Object.entries(resultMap.scopes).forEach(([k, v]) => {
+      expect(v).not.toBe(map.scopes[k]);
+    });
+  });
+
+  it('should compose with the empty map on the right', () => {
+    const map = parseFromString(JSON.stringify({
+      imports: { 'https://a/': ['https://b/'] },
+      scopes: {
+        'https://c/': { 'https://d/': ['https://e/'] }
+      }
+    }), mapBaseURL);
+
+    const resultMap = appendMap(map, parseFromString('{}', mapBaseURL));
+
+    expect(resultMap).toStrictEqual(map);
+    expect(resultMap.imports).not.toBe(map.imports);
+    Object.entries(resultMap.imports).forEach(([k, v]) => {
+      expect(v).not.toBe(map.imports[k]);
+    });
+    expect(resultMap.scopes).not.toBe(map.scopes);
+    Object.entries(resultMap.scopes).forEach(([k, v]) => {
+      expect(v).not.toBe(map.scopes[k]);
+    });
+  });
+
+  it('should compose maps that do not interact in any way', () => {
+    expect(composeMaps([
+      {
+        imports: { 'https://a/': 'https://b/' }
+      },
+      {
+        imports: { 'https://c/': 'https://d/' }
+      },
+      {
+        imports: { 'https://e/': 'https://f/' }
+      }
+    ])).toStrictEqual({
+      imports: {
+        'https://a/': ['https://b/'],
+        'https://c/': ['https://d/'],
+        'https://e/': ['https://f/']
+      },
+      scopes: {}
+    });
+  });
+
+  it('should compose maps that interact via cascading', () => {
```

Comment:
Can you add some composition tests with relative URLs, specifically ones that are different representations of each other? For example, both /b and ../b refer to the same URL, and so if one map uses the former and another the latter, you should still get this kind of cascading behavior.

This should probably be done for the next several tests, at least until you feel confident we've covered the potential places that this could show up.